services/controller/src/plz/controller/instances/instance_base.py

```diff
@@ -330,49 +330,108 @@ def release_instance(
         :param instance_ids: instances to dispose of. A value of `None` means
-               all instances in the group
+               all instances in the group running jobs of the user
+        :param ignore_ownership: kill instances even if they're running jobs
+               of other users
+        :param including_idle: when killing all user instances, kill idle ones
         :param force_if_not_idle: force termination for non-idle instances
         :raises: :class:`ProviderKillingInstancesException` if some instances
                  failed to terminate
         :raises: :class:`NoInstancesFound` if asked for the termination of all
                  instances, and there are no instances
         """
-        terminate_all_instances = instance_ids is None
+        terminate_all_user_instances = instance_ids is None
 
-        if not terminate_all_instances:
+        if not terminate_all_user_instances:
             unprocessed_instance_ids = [i for i in instance_ids]
         else:
             unprocessed_instance_ids = []
 
         instance_ids_to_messages = {}
         there_is_one_instance = False
         for instance in self.instance_iterator(only_running=False):
-            if instance.is_terminated():
+            if self._must_skip_killing_instance(
+                    ignore_ownership,
+                    including_idle,
+                    instance,
+                    instance_ids,
+                    instance_ids_to_messages,
+                    terminate_all_user_instances,
+                    unprocessed_instance_ids,
+                    user):
                 continue
+
             there_is_one_instance = True
-            if terminate_all_instances or instance.instance_id in instance_ids:
-                    try:
-                        instance.kill(force_if_not_idle)
-                    except KillingInstanceException as e:
-                        instance_ids_to_messages[instance.instance_id] = \
-                            e.message
-                    if not terminate_all_instances:
-                        unprocessed_instance_ids.remove(instance.instance_id)
+            try:
+                instance.kill(force_if_not_idle)
+            except KillingInstanceException as e:
+                instance_ids_to_messages[instance.instance_id] = \
+                    e.message
+            if not terminate_all_user_instances:
+                unprocessed_instance_ids.remove(instance.instance_id)
 
         for instance_id in unprocessed_instance_ids:
             instance_ids_to_messages[instance_id] = 'Instance not found'
 
         if len(instance_ids_to_messages) > 0:
             raise ProviderKillingInstancesException(instance_ids_to_messages)
 
-        if terminate_all_instances and not there_is_one_instance:
+        if terminate_all_user_instances and not there_is_one_instance:
             raise NoInstancesFoundException()
 
+    def _must_skip_killing_instance(
```

Comment:
Just a thought: something like `skip_killing = False` feels like a double negative. The method would be easier to read if it was called '_must_kill_instance' instead. Better to kill if you must than to skip killing if you mustn't...

Reply:
Good suggestion, done